### PR TITLE
lib/vfscore: Implement individual volume automounting

### PR DESCRIPTION
**Pull request by mogasergiu**
Provide a command-line argument `vfs.fstab` that is meant to contain a list of whitespace separated strings with the following format:
```
	vfs.fstab=[
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		...
	]
```

The core will parse the provided strings as volume information and attempt to mount them accordingly.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]

### CODE REVIEW

File: lib/vfscore/Makefile.uk
Comment (skuenzer) [inline lib/vfscore/Makefile.uk]: Can we name the file `automount.c`? I think as a little clean-up it makes sense to rename `rootfs.c` and then add your code on top. Then we have it within one file.

File: lib/vfscore/volumes.c
Comment (skuenzer) [inline lib/vfscore/volumes.c]: Can we make this available via menu config? Also make it configurable if automount is available?

File: lib/vfscore/volumes.c
Comment (skuenzer) [inline lib/vfscore/volumes.c]: great! It is a bit clearer maybe if we have the registrations in the same file. We could also include your code to the rootfs mount code.

File: lib/vfscore/Makefile.uk
Comment (mogasergiu) [reply lib/vfscore/Makefile.uk]: Are you sure? Because that would make volumes depend on `CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS`. And there is a case where you may want to have atuomatic mount of volumes without having the `rootfs.c` functionality.

Comment (skuenzer) [reply lib/vfscore/Makefile.uk]: Both true but you could just guard the ROOTFS code in the file and keep the source file part of the build, independent of the switch, right? My point is that it is easier to lookup any automount related code and that it is close together from an organization perspective.

Comment (mogasergiu) [reply lib/vfscore/Makefile.uk]: So guard two thirds of the file? Does it not look like bad design since the separation is so big to not have them in separate files so that the separation is transparent from the source code point of view?

**Review by StefanJum — COMMENTED** [review]
This seems to work fine @mogasergiu, I'll add the tag after you resolve Simon's comments.

File: lib/vfscore/volumes.c
Comment (RaduNichita) [inline lib/vfscore/volumes.c]:
```suggestion
 *		"<source_device>:<mount_target>:<fs_driver>:<options>",
```
I think there should be a comma here.

File: lib/vfscore/volumes.c
Comment (RaduNichita) [inline lib/vfscore/volumes.c]: Could you elaborate a little bit about priorities. Why it should be 4 here?

**Review by RaduNichita — CHANGES_REQUESTED** [review]
Great work, @mogasergiu :muscle: . I left some minor nitpicks.
I will approve after that!

File: lib/vfscore/volumes.c
Comment (mogasergiu) [reply lib/vfscore/volumes.c]: @skuenzer AFAIK, the `uklibparam` design makes use of whitespaces in such cases :thinking:.

File: lib/vfscore/volumes.c
Comment (mogasergiu) [reply lib/vfscore/volumes.c]: What I meant was that `rootfs.c` has a `uk_rootfs_initcall_prio(vfscore_rootfs, 4);`. Therefore, in order to ensure this can only be called after an explicit rootfs, give this a priority of `4 + 1 = 5`. I was referring to the `rootfs`. I will make the comment more clear :100:.

File: lib/vfscore/volumes.c
Comment (skuenzer) [reply lib/vfscore/volumes.c]: Yes, no commas. This has actually a pragmatic reason because that missing comma saves us parsing code. We can just rely on `ukargparse` for doing the argument split and array parsing. `ukargparse` also handles quotes and escaped quotes correctly. `uklibparam` automatically benefits from this.

File: lib/vfscore/volumes.c
Comment (skuenzer) [reply lib/vfscore/volumes.c]: I think if the `uk_rootfs_initcall_prio()` calls are next to each other (I am speaking about merged `rootfs.c` and `volumes.c`), then this is also clear. Additionally, `<uk/prio.h>` provides some computation macros so that this could be by using those also be a bit more expressive by the code. And the last alternative I see is to use only a single automount function that handles first `rootfs` and then `volumes`...

**Review by StefanJum — COMMENTED** [review]
I retested this and everything works fine, thanks!
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

File: lib/vfscore/automount.c
Comment (skuenzer) [inline lib/vfscore/automount.c]: This is more a minor detail, more aesthetically: I'd prefer two functions at the same level:
`vfscore_automount_rootfs`
`vfscore_automount_volumes`
and a third function that calls these two and is registered to the init table.
btw, it is a clever idea to create empty function bodies depending on the configuration.

**Review by RaduNichita — APPROVED** [review]
All good now. Thanks, @mogasergiu :1st_place_medal: 

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

File: lib/vfscore/Config.uk
Comment (skuenzer) [inline lib/vfscore/Config.uk]: Can we also introduce an option that switches this on or off? I am thinking of something along these lines:
```
config LIBVFSCORE_AUTOMOUNT_VOLUMES
bool "Automatically mount list of volumes from kernel arguments"
selects LIBUKLIBPARAM
default n

if LIBVFSCORE_AUTOMOUNT_VOLUMES
    config LIBVFSCORE_AUTOMOUNT_VOLUMES_COUNT
    # ...
endif
```

**Review by StefanJum — COMMENTED** [review]
Sill no bugs found in the wild :sunglasses: 
The only thing I would do is move the `help` section from under the `LIBVFSCORE_MAX_VOLUMES_COUNT` to the `LIBVFSCORE_AUTOMOUNT_VOLUMES` option. It's not really related to the max number of volumes.

**Review by razvanvirtan — APPROVED** [review]
All good for me, thanks!
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

**Comment by mogasergiu** [comment]
> Sill no bugs found in the wild sunglasses The only thing I would do is move the `help` section from under the `LIBVFSCORE_MAX_VOLUMES_COUNT` to the `LIBVFSCORE_AUTOMOUNT_VOLUMES` option. It's not really related to the max number of volumes.

Good point, I simply copy-pasted. :rofl:

**Review by StefanJum — APPROVED** [review]
All good now
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

File: lib/vfscore/automount.c
Comment (mogasergiu) [inline lib/vfscore/automount.c]: Oops should have been `CONFIG_LIBUKCPIO`

Comment (skuenzer) [reply lib/vfscore/automount.c]: Probably we should just fall back to a default message, something like saying: `Unknown filesystem: initrd`. I guess we have something like this already built-in with `mount()`. So, I think a separate handler for "initrd-support-is-not-present" is not needed.

File: lib/vfscore/automount.c
Comment (skuenzer) [inline lib/vfscore/automount.c]: What is `vv.sdev` saying in the initrd case? Maybe you could place this message under the else case and print a message for initrd like: "Extracting initrd to %s...\n"

Comment (mogasergiu) [reply lib/vfscore/automount.c]: `vv.sdev` is a mandatory argument as per `fstab` syntax:

```
vfs.fstab=[
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		...
	]
```
`vv.sdev` is therefore `<src_dev>`.

EDIT: The initrd specific printing is already done by `vfscore_mount_initrd_volume`

File: lib/vfscore/automount.c
Comment (mogasergiu) [reply lib/vfscore/automount.c]: Right then, I will just let this be an empty stub.

File: lib/vfscore/automount.c
Comment (skuenzer) [inline lib/vfscore/automount.c]: I think this message is not needed anymore and it is incorrect. initrd can go now to any submountpoint, right?

File: lib/vfscore/automount.c
Comment (skuenzer) [inline lib/vfscore/automount.c:166]: I think these spaces between message and return code is not necessary for these small if blocks. I leave it up to you if you want to keep them or not.

File: lib/vfscore/automount.c
Comment (skuenzer) [reply lib/vfscore/automount.c]: My question was that the initrd mounting code does not need a source device as parameter. According to the code we accept anything that is in there which is actually fine. :+1:

File: lib/vfscore/automount.c
Comment (mogasergiu) [reply lib/vfscore/automount.c]: Now that you mention it, [mount](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/mount.c#L122) already has a similar message. Therefore the message should be dropped from `vfscore_automount_fstab_volumes` as well :+1:.

**Comment by skuenzer** [comment]
I just noticed, that in your second commit that removes the `vfs.roofs` parameter, etc. you also need to update `Config.uk`:
```diff
 config LIBVFSCORE_AUTOMOUNT_ROOTFS
 bool "Automatically mount a root filesysytem (/)"
 default n
-imply LIBUKLIBPARAM
 help
-	Automatically mounts '/' during boot. If `libuklibparam` is
-	compiled in, the default root filesystem and mount options can
-	be changed with the following library parameters:
-	'vfs.rootfs', 'vfs.rootdev', 'vfs.rootflags', and 'vfs.rootopts'
+	Automatically mounts '/' during boot. The parameters are
+	compiled-in and cannot be changed at runtime.
```

This is a minor comment: Instead of keeping the parameters set from config options from line 76 to 98, you could directly fill the struct in line 226 with those.

I aslo jsut notice that you are using `#ifdef CONFIG_ABC` although you can use `#if CONFIG_ABC` which is slightly more powerful because it catches the case when you manually do `#define CONFIG_ABC 0`. If you use `#if` on an undefined pre-processor macro, it will return false.

File: lib/vfscore/automount.c
Comment (skuenzer) [inline lib/vfscore/automount.c:267]: We should propagate failure as soon as one mount operation failed

**Comment by mogasergiu** [comment]
> I aslo jsut notice that you are using `#ifdef CONFIG_ABC` although you can use `#if CONFIG_ABC` which is slightly more powerful because it catches the case when you manually do `#define CONFIG_ABC 0`. If you use `#if` on an undefined pre-processor macro, it will return false.

Is this just general advice? The problem is if you do `#if` on a string you are going to be met with a preprocessor error. So I think I will stick to `#ifdef`'s in the `CONFIG_LIBVFSCORE_ROOT*` macro's :thinking:.

**Review by skuenzer — APPROVED** [review]
Thanks a lot for the work!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request closed without merging**
